Design/Instrument-2.0/ScanningInstruments_Design.md

```diff
@@ -0,0 +1,170 @@
+
+The `timeIndex` argument is optional, so algorithms that do not care about time indexing will simply return the first/only time index. `DetectorInfo` should also contain a vector of start and end times for each time index. This will mean any information stored as a time series sample log can also be obtained for the correct step, for example the monitor counts for normalisation.
+
+`DetectorInfo` will ultimately store the information for positions and rotation, so will need to do this in a way that can be accessed by both the detector index and time index. Masking would be done at the spectrum level, if it is required for a detector for some time steps, but not all. `DetectorInfo` will also hold the start and end times for each step in the scan. The objects it will need to hold for this would be:
+* `std::vector<std::vector<Kernel::V3D>> positions`
+* `std::vector<std::vector<Kernel::V3D>> rotations`
+* `std::vector<std::vector<std:pair<Kernel::DateAndTime, Kernel::DateAndTime>>> timeRanges`
+
+Appropriate typedefs should be used here for readability. All of the vector object would have the same size, the number of steps in the scan for the vector of vectors, and the number of detectors for the vector of positions.
+
+#### `SpectrumInfo`
+
+Currently [`SpectrumInfo`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/SpectrumInfo.h) has a private `getDetector` method for finding the detector for the workspace. This uses the `getSpectrum(index)` method on `MatrixWorkspace` to return the detector IDs for the spectrum.
+
+A new method should be added, for example `std::set<std::pair<detid_t, size_t>> getTimeIndexedDetectorIDs(index)`. The time indices for each detector can be stored in `SpectrumInfo` as a `std::set<std::pair<detid_t, size_t>>` within a vector of the same size as the number of workspace spectra. This will avoid the need to change methods on `MatrixWorkspace` or `ISpectrum/Spectrum` to return the set of detector IDs and time indexes for the spectrum. This is something that will help to move away from the incremental construction of `DetectorInfo`.
+
+This can be used to access methods on `DetectorInfo`, for example `m_detectorInfo.twoTheta(detIndex, timeIndex)`.
+
+#### `MatrixWorkspace` & `ISpectrum`/`Spectrum`
+
+The original methods such as `this->detectorIDs.insert(detID)` should remain for now. This allows existing algorithms, e.g. loaders, that do not need time indexing to work as before. Similarly the `getDetectorIDs(index)` method would still return detector IDs without the time indexing information.
+
+**Note:** working directly with the detector IDs is something that should be eliminated with the use of `SpectrumInfo` and `DetectorInfo`. The use of a `getDetectorIDs` should not be needed in the final Instrument 2.0. Similarly the method `getSpectrum(index)` on `MatrixWorkspace` should eventually be eliminated.
+
+#### StepScan Construction
```

Comment:
I have to think a bit about how this will work while we are still wrapping the old instrument. In particular construction order of workspace and instrument, etc.